File: Python/PASTRY.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-

import argparse, time
import numpy as np
import datatable as dt
import scipy.stats as stats
from numpy import savetxt

def pastry(indir, maf, outdir, k, ncase, ncont, ncontshared):

	# K : Combining Studies (# of files)
	# X : Genotype of K study
	# Y : Phenotype of k study
	# nsnp : # of SNP for one study

	# Importing input files and setting parameters
	print("Your input files count for meta-analysis : ",k)

	casemaf=dt.fread(maf+"casemaf.out").to_numpy()
	contmaf=dt.fread(maf+"contmaf.out").to_numpy()
	nsnp=len(casemaf[:,0])
	#maf=dt.fread(mafpath).to_pandas() # for (.frq.cc files)
	#casemaf=np.array(maf["MAF_A"]) # for (.frq.cc files)
	#contmaf=np.array(maf["MAF_U"]) # for (.frq.cc files)
	unit=np.repeat(1,k).reshape(1,k)
	unit_t=np.transpose(unit)

	nstudy=np.repeat(ncont+ncase,k)
	ncont=np.repeat(ncont,k)
	ncase=np.repeat(ncase,k)

	ls_stat=np.zeros(nsnp)
	pastry_stat=np.zeros(nsnp)

	betas=np.zeros((k,nsnp))
	ses=np.zeros((k,nsnp))

	for i in range(1,k+1):
		study=dt.fread(indir+str(i)+".out").to_numpy()
		betas[i-1]=study[:,0]
		ses[i-1]=study[:,1]

	# Na handling
	betas=np.nan_to_num(betas)
	ses=np.nan_to_num(ses)

	# LS
	## Generating Correlation Matrix
	ls_R=np.diagflat(np.repeat(1,k)).astype(float)
	pastry_R=np.diagflat(np.repeat(1,k)).astype(float)
	
	for i in range(k):
		for j in range(k):
			if i!=j:
				num=ncontshared*np.sqrt((ncase[i]*ncase[j])/(ncont[i]*ncont[j])) #numerator
				den=np.sqrt(nstudy[i]*nstudy[i]) #denominator
				ls_R[i,j]=num/den
	
	inv_ls_R=np.linalg.inv(ls_R)


	invstdmat={}

	ls_beta=np.zeros(nsnp)
	ls_stat=np.zeros(nsnp)

	pastry_beta=np.zeros(nsnp)
	pastry_stat=np.zeros(nsnp)

	## Generate Statistics
	for n in range(nsnp):
		p0cont=1-contmaf[n]
		p1cont=contmaf[n]

		p0case=1-casemaf[n]
		p1case=casemaf[n]

		invstdmat[n]=np.linalg.pinv(np.diagflat(ses[:,n]))
		ls_invsigma=np.dot(np.dot(invstdmat[n],inv_ls_R),invstdmat[n])
		ls_invvar=np.dot(np.dot(unit,ls_invsigma),unit_t)
		ls_beta_num=np.dot(np.dot(unit,ls_invsigma),betas[:,n])

		ls_beta[n]=ls_beta_num/ls_invvar
		ls_var=1/ls_invvar

		ls_stat[n]=ls_beta[n]/np.sqrt(ls_var)

	# PASTRY
	start = time.time()
	for n in range(nsnp):
		for i in range(k):
			for j in range(k):
				if i!=j:
					p0cont=1-contmaf[n]
					p1cont=contmaf[n]

					p0case=1-casemaf[n]
					p1case=casemaf[n]

					ntotali=ncont[i]+ncase[i]
					ntotalj=ncont[j]+ncase[j]

					exp_alpha_i = np.exp(np.log(ncase[i]/(ncont[i]))-(ls_beta[n]*(((ncase[i]*casemaf[n])+(ncont[i]*contmaf[n]))/ntotali)))
					exp_alpha_j = np.exp(np.log(ncase[j]/(ncont[j]))-(ls_beta[n]*(((ncase[j]*casemaf[n])+(ncont[j]*contmaf[n]))/ntotalj)))

					x0_i = exp_alpha_i
					x1_i = exp_alpha_i*np.exp(ls_beta[n])

					x0_j = exp_alpha_j
					x1_j = exp_alpha_j*np.exp(ls_beta[n])

					## Information Matrix
					x0=np.array([1,0,0,0]).reshape(2,2)
					x1=np.array([1,1,1,1]).reshape(2,2)

					up_covmat=np.array([0,0,0,0]).reshape(2,2)

					info_i=2*ncase[i]*(p0case*(x0_i/(1+x0_i)**2)*x0 + p1case*(x1_i/(1+x1_i)**2)*x1) + 2*(ncont[i])*(p0cont*(x0_i/(1+x0_i)**2)*x0 + p1cont*(x1_i/(1+x1_i)**2)*x1)
					info_j=2*ncase[j]*(p0case*(x0_j/(1+x0_j)**2)*x0 + p1case*(x1_j/(1+x1_j)**2)*x1) + 2*(ncont[j])*(p0cont*(x0_j/(1+x0_j)**2)*x0 + p1cont*(x1_j/(1+x1_j)**2)*x1)

					## Covarinace Matrix
					#cov00=p0cont*(x0_i*x0_j/(((1+x0_i)*(1+x0_j))**2))*x0
					#cov11=p1cont*(x1_i*x1_j/(((1+x1_i)*(1+x1_j))**2))*x1

					cov00=p0cont*(x0_i*x0_j/((1+x0_i)*(1+x0_j)))*x0
					cov11=p1cont*(x1_i*x1_j/((1+x1_i)*(1+x1_j)))*x1

					covmatrix=2*ncontshared*(cov00+cov11)

					inv_i = np.linalg.pinv(info_i)
					inv_j = np.linalg.pinv(info_j)

					covfinal=np.dot(np.dot(inv_i,covmatrix),inv_j)

					### Correlation matrix for PASTRY
					cor=covfinal[1,1]/np.sqrt(np.dot(inv_i[1,1],inv_j[1,1]))
					pastry_R[i,j]=cor
					pastry_R[j,i]=cor

		pastry_invsigma=np.dot(np.dot(invstdmat[n],np.linalg.inv(pastry_R)),invstdmat[n])
		pastry_invvar=np.dot(np.dot(unit,pastry_invsigma),unit_t)

		pastry_var=1/pastry_invvar
		beta_num=np.dot(np.dot(unit,pastry_invsigma),betas[:,n])

		pastry_beta[n]=beta_num/pastry_invvar
		pastry_stat[n]=pastry_beta[n]/np.sqrt(pastry_var)

	end = time.time()
	print("Your output is writing...\n")
	savetxt("{}_PASTRY_zstat.py.out".format(outdir),pastry_stat,fmt='%.3f',header="PASTRY_Z",delimiter="\n")
	savetxt("{}_LS_zstat.py.out".format(outdir),ls_stat,fmt='%.3f',header="LS_Z",delimiter="\n")
	#savetxt("{}_contmaf.py.out".format(outdir),contmaf,delimiter="\n") # for (.frq.cc files)
	#savetxt("{}_casemaf.py.out".format(outdir),casemaf,delimiter="\n") # for (.frq.cc files)
	print("Done.\n")
	print(" * PASTRY processed time : "+str(end - start))

	return()

# PASTRY module run as the main program.
if __name__ == "__main__":


	print("\n")
	print("================ PASTRY v 1.0 ================ \n")
	print("==== v.1.0 Seoul National University Hanlab ==== \n")
	print("\n")

	parser = argparse.ArgumentParser()
	parser.set_defaults(method = pastry)
	parser.add_argument("-i", "--input", type=str,help=":input files prefix name (.out)") # input files prefix
	parser.add_argument("-m", "--maf", type=str, default="./",help=":maf files directory (casemaf.out, contmaf.out) ") # input maf file directory path          
	parser.add_argument("-o", "--out", type=str, help=":output file prefix name (.out)") # output files prefix
	parser.add_argument("-n","--N", type=int,help=":# of studies combined", metavar="integer") # meta-analysis studies
	parser.add_argument("-s", "--shared", type=int,help=":# of shared control", metavar="integer")  # shared control N
	parser.add_argument("-e","--case", type=int,help=":# of case", metavar="integer")            # case N
	parser.add_argument("-l","--cont", type=int,help=":# of control", metavar="integer")          # control N
	options = parser.parse_args()

	if options.input is None:
		raise ValueError('input file is required. -h for help message.')
	if options.maf is None:
		raise ValueError('frequency file is required. -h for help message.')
	if options.out is None:
		raise ValueError('Result file prefix is required. -h for help message.')
	if options.N is None:
		raise ValueError('# N is required. -h for help message.')
	if options.shared is None:
		raise ValueError('# of shared is required. -h for help message.')
	if options.case is None:
		raise ValueError('# of control is required. -h for help message.')
	if options.cont is None:
		raise ValueError('# of control is required. -h for help message.')

	pastry(indir=options.input, maf=options.maf, outdir=options.out, k=options.N, ncase=options.case, ncont=options.cont, ncontshared=options.shared)
	print("PASTRY Analysis & Writing Results Done. ========= ")
	print("================================================= \n ")

File: R/PASTRY.R
#!/usr/bin/env Rscript

## Now load or install & load all
packages = c("data.table", "optparse", "pracma")

package.load <- lapply(packages,
  FUN = function(x) {
    if (!require(x, character.only = TRUE)) {
      install.packages(x, dependencies = TRUE)
      suppressMessages(library(x,character.only=TRUE, quietly = T))
      }
  }
)

## Option argparse
option_list = list(
    make_option(c("-i", "--input"), type="character", default=NULL,
              help=":input files prefix name (.out)", metavar="character"), # input files prefix
    make_option(c("-m", "--maf"), type="character", default="./",
              help=":maf files directory (casemaf.out, contmaf.out)", metavar="character"), # input maf file directory path          
    make_option(c("-o", "--out"), type="character", default=NULL, 
              help=":output file prefix name (.out)", metavar="character"), # output files prefix
    make_option(c("-n","--N"), type="integer", default=NULL,
              help=":# of studies combined", metavar="integer"), # meta-analysis studies
    make_option(c("-s", "--shared"), type="integer", default=NULL,
              help=":# of shared control", metavar="integer"),  # shared control N
    make_option(c("-l","--cont"), type="integer", default=NULL,
              help=":# of control", metavar="integer"),          # control N
    make_option(c("-e","--case"), type="integer", default=NULL,
              help=":# of case", metavar="integer")            # case N
);


opt_parser = OptionParser(option_list=option_list);
opt = parse_args(opt_parser);

if ( length(opt$input) == 0 | length(opt$N) == 0 | length(opt$shared) == 0 | length(opt$case) == 0 | length(opt$cont) == 0) {
    stop("Please provide PASTRY parameters (check the required parameters using 'Rscript PASTRY.R -h')")
}

## Function define
pastry <- function(files=opt$input,maf=opt$maf,outdir=opt$out,N=opt$N, ncase=opt$case, ncont=opt$cont, ncontshared=opt$shared){

        files <- paste(rep(files,N),seq(1,N),sep="")
        
        casemaf <- as.matrix(read.table(paste(maf,"casemaf.out",sep=""),header=T))
        contmaf <- as.matrix(read.table(paste(maf,"contmaf.out",sep=""),header=T))

        #maf=as.matrix(fread(maf,header=T)) # for (.frq.cc files)
        #casemaf=as.numeric(maf[,"MAF_A"]) # for (.frq.cc files)
        #contmaf=as.numeric(maf[,"MAF_U"]) # for (.frq.cc files)

        ncase <- rep(ncase,N)
        ncont <- rep(ncont,N)
        ncontspe <- ncont-ncontshared
        casenshared <- 0  # only consider overlapping controls

        # Load File names and snp #
        cat(paste("You have ",N," Input files. \n", sep=""))

        study <- as.matrix(fread(paste(files[1],".out",sep="")))
        nsnp <- length(study[,1])

       cat(paste("You have ",nsnp," SNPs in each studies. \n \n", sep="")) 

        # Matrix Initialize
        betas <- matrix(0,N,nsnp)
        ses <- matrix(0,N,nsnp)
        ls_stat <- rep(0,nsnp)
        pastry_stat <- rep(0,nsnp)
        unit <- t(rep(1,N)) ## e^t = unit

        betas[1,] <- as.numeric(study[,1])
        ses[1,] <- as.numeric(study[,2])

        ls_stat <- rep(0,nsnp)
        pastry_stat <- rep(0,nsnp)

        # Beta, SE Files loading
        for (l in 2:N){
        study <- as.matrix(fread(paste(files[l],".out",sep="")))
        betas[l,] <- as.numeric(study[,1])
        ses[l,] <- as.numeric(study[,2])
        }

        ## NA handling
        betas[is.na(betas)] <- 0
        ses[is.na(ses)] <- 0

        ## Lin_correlation matrix
        R <- diag(N)
        for (i in 1:N) {
            for (j in 1:N) {
                if (i == j) {next}
                else {
                    first <- ncontshared*sqrt(((ncase[i]*ncase[j]))/((ncontspe[i]+ncontshared)*(ncontspe[j]+ncontshared)))
                    sec <- casenshared*sqrt(((ncontspe[i]+ncontshared)*(ncontspe[j]+ncontshared))/((ncase[i])*(ncase[j])))
                    rij <- first/sqrt((ncase[i]+ncontspe[i]+ncontshared)*(ncase[j]+ncontspe[j]+ncontshared))
                    R[i,j] = R[j,i] = rij
                }
            }
        }

        invR <- solve(R)
        cat(" * Lin Correlation Matrix generated. \n")

        ## Generating Statistics for LS & PASTRY
        for (k in 1:nsnp){
            for (i in 1:N){
                for (j in 1:N){
                    if (i != j){

                        ntotali <- ncont[i]+ncase[i]
                        ntotalj <- ncont[j]+ncase[j]

                        p0cont <- 1-contmaf[k]
                        p1cont <- contmaf[k]

                        p0case <- 1-casemaf[k]
                        p1case <- casemaf[k]

                        x_0 <- matrix(c(1,0,0,0),nrow=2,ncol=2)
                        x_1 <- matrix(c(1,1,1,1),nrow=2,ncol=2)

                        info_i <- matrix(c(0,0,0,0),nrow=2,ncol=2)
                        info_j <- matrix(c(0,0,0,0),nrow=2,ncol=2)

                        up_covmat <- matrix(c(0,0,0,0),nrow=2,ncol=2)

                        ## LS method
                        if (k == 1 && i == 1 && j == 2){cat(" * LS statistics are generating... for each SNP. \n")}
                        invstdmat <- pinv(diag(ses[,k]))
                        invsigma <- invstdmat %*% invR %*% invstdmat
                        invvar <- unit %*% invsigma %*% t(unit)

                        beta_num <- unit %*% invsigma %*% (betas[,k])

                        ls_beta <- beta_num/invvar
                        ls_var <- 1/invvar

                        ls_stat[k] <- ls_beta/sqrt(ls_var)

                        exp_alpha_i <- as.numeric(exp(log(ncase[i]/(ncontspe[i]+ncontshared)) - (ls_beta*((ncase[i]*casemaf[k]+(ncontspe[i]+ncontshared)*contmaf[k])/ntotali))))
                        exp_alpha_j <- as.numeric(exp(log(ncase[j]/(ncontspe[j]+ncontshared)) - (ls_beta*((ncase[j]*casemaf[k]+(ncontspe[j]+ncontshared)*contmaf[k])/ntotalj))))

                        x0i <- exp_alpha_i
                        x1i <- as.numeric(exp_alpha_i*exp(ls_beta))

                        x0j <- exp_alpha_j
                        x1j <- as.numeric(exp_alpha_j*exp(ls_beta))

                        if (k == 1 && i == 1 && j == 2) {cat(" * PASTRY statistics are generating... for each SNP. \n")
                        start_time <- Sys.time()}

                        ### Information Matrix for PASTRY
                        info_i <- 2*ncase[i]*(p0case*(x0i/(1+x0i)**2)*x_0 + p1case*(x1i/(1+x1i)**2)*x_1) + 2*(ncontspe[i]+ncontshared)*(p0cont*(x0i/(1+x0i)**2)*x_0 + p1cont*(x1i/(1+x1i)**2)*x_1)
                        info_j <- 2*ncase[j]*(p0case*(x0j/(1+x0j)**2)*x_0 + p1case*(x1j/(1+x1j)**2)*x_1) + 2*(ncontspe[j]+ncontshared)*(p0cont*(x0j/(1+x0j)**2)*x_0 + p1cont*(x1j/(1+x1j)**2)*x_1)

                        ### Covarinace Matrix for PASTRY
                        cov00 <- p0cont*(x0i*x0j/((1+x0i)*(1+x0j)))*x_0
                        cov11 <- p1cont*(x1i*x1j/((1+x1i)*(1+x1j)))*x_1

                        covmatrix <- 2*ncontshared*(cov00+cov11)

                        invi <- pinv(info_i)
                        invj <- pinv(info_j)
                        covfinal <- invi %*% covmatrix %*% invj
                    }
                }
            }
            
        ### Correlation matrix for PASTRY
        cor <- (covfinal[2,2]/sqrt(invi[2,2]*invj[2,2]))
        PASTRY.cor <- matrix(rep(cor,N*N),nrow=N,ncol=N)
        diag(PASTRY.cor) <- 1

        if (k == 1 && i == 1 && j == 2){print(PASTRY.cor)}
        invsigmaPASTRY <- invstdmat %*% solve(PASTRY.cor) %*% invstdmat
        invvarpastry <- unit %*% invsigmaPASTRY %*% t(unit)

        pastry_var <- 1/invvarpastry
        beta_num <- unit %*% invsigmaPASTRY %*% (betas[,k])

        pastry_beta <- beta_num/invvarpastry
        pastry_stat[k] <- pastry_beta/sqrt(pastry_var)
    } 
    end_time <- Sys.time()

out="PASTRY_"
cat("\n\nAll done. Writing your results... \n")
cat("Your PASTRY output files are generated. \n")
fwrite(format(as.data.frame(ls_stat),digits=3),paste(outdir,"_LS_zstat.r.out",sep=""),sep="\n",row.names=FALSE)
fwrite(format(as.data.frame(pastry_stat),digits=3),paste(outdir,"_PASTRY_zstat.r.out",sep=""),sep="\n",row.names=FALSE)
#fwrite(as.data.frame(casemaf),paste(outdir,"_",out,"casemaf.out",sep=""),sep="\n",row.names=FALSE) # for (.frq.cc files)
#fwrite(as.data.frame(contmaf),paste(outdir,"_",out,"contmaf.out",sep=""),sep="\n",row.names=FALSE) # for (.frq.cc files)
cat("\n")
cat(paste("PASTRY running time in R : ", end_time - start_time, ". \n", sep=""))
}


## Main

cat("\n")
cat("================ PASTRY v 1.0 ================ \n")
cat("==== v.1.0 Seoul National University Hanlab ==== \n")
cat("\n")

pastry()

cat("================ PASTRY END ================")
cat("\n\n")